gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/keyedgrouping/functions/BuildTuplesFromElements.java

```diff
@@ -0,0 +1,137 @@
+ * aggregated.
+ * <p>
+ * <i>Note: </i> This function sets all grouping keys and aggregate values, make sure to set
+ * additional
+ * fields, if {@code tupleDataOffset} is not {@code 0}.
+ *
+ * @param <E> The element type.
+ */
+public class BuildTuplesFromElements<E extends Element>
+  implements MapFunction<E, Tuple>, ResultTypeQueryable<Tuple> {
+
+  /**
+   * The grouping key functions.
+   */
+  private final List<KeyFunction<E, ?>> keys;
+
+  /**
+   * The aggregate functions.
+   */
+  private final List<AggregateFunction> aggregateFunctions;
+
+  /**
+   * The number of fields to be reserved for IDs.
+   * Those fields will be of type {@link GradoopId}.
+   */
+  private final int tupleDataOffset;
+
+  /**
+   * The types of the produced tuple.
+   */
+  private final TypeInformation<?>[] elementTypes;
+
+  /**
+   * Reduce object instantiations.
+   */
+  private final Tuple reuseTuple;
+
+  /**
+   * Initialize this function, setting the grouping keys and aggregate functions.
+   *
+   * @param tupleDataOffset    The number of tuple fields reserved for IDs.
+   * @param keys               The grouping keys.
+   * @param aggregateFunctions The aggregate functions used to determine the aggregate property
+   */
+  public BuildTuplesFromElements(int tupleDataOffset, List<KeyFunction<E, ?>> keys,
+    List<AggregateFunction> aggregateFunctions) {
+    this.tupleDataOffset = tupleDataOffset;
+    if (tupleDataOffset < 0) {
+      throw new IllegalArgumentException(
+        "The number of reserved tuple fields must not be negative.");
+    }
+    this.keys = Objects.requireNonNull(keys);
+    this.aggregateFunctions = Objects.requireNonNull(aggregateFunctions);
+    final int tupleSize = tupleDataOffset + keys.size() + aggregateFunctions.size();
+    if (tupleSize > Tuple.MAX_ARITY) {
+      throw new UnsupportedOperationException("Number of elements is too high for tuple: " +
+        tupleSize);
+    }
+    elementTypes = new TypeInformation[tupleSize];
+    for (int i = 0; i < tupleDataOffset; i++) {
+      elementTypes[i] = TypeInformation.of(GradoopId.class);
+    }
+    // Fill grouping key types.
+    for (int i = 0; i < keys.size(); i++) {
+      elementTypes[i + tupleDataOffset] = keys.get(i).getType();
+    }
+    // Fill remaining spots with property value types.
+    for (int i = 0; i < aggregateFunctions.size(); i++) {
+      elementTypes[i + keys.size() + tupleDataOffset] = TypeInformation.of(PropertyValue.class);
+    }
+    reuseTuple = Tuple.newInstance(tupleSize);
+    // Fill first fields with default ID values.
+    for (int i = 0; i < tupleDataOffset; i++) {
+      reuseTuple.setField(GradoopId.NULL_VALUE, i);
+    }
+  }
+
+  @Override
+  public TypeInformation<Tuple> getProducedType() {
+    TypeInformation<?>[] componentTypes = new TypeInformation[elementTypes.length];
+    for (int i = 0; i < componentTypes.length; i++) {
+      componentTypes[i] = elementTypes[i];
```

Comment:
You can use `Arrays.copyOf()`. Why do you have to copy this at all?

Reply:
This only became unnecessary after some refactoring, I'll remove it.